src/charm.py

```diff
@@ -738,10 +841,105 @@ def _get_initialization_secret_from_peer_relation(self) -> Tuple[str, List[str]]
         except (TypeError, SecretNotFoundError, AttributeError):
             raise PeerSecretError(secret_name=VAULT_INITIALIZATION_SECRET_LABEL)
 
+    def _create_s3_session(self, s3_parameters: dict) -> Optional[boto3.session.Session]:
+        """Creates S3 session.
+
+        Args:
+            s3_parameters: Dictionary of the S3 parameters.
+
+        Returns:
+            boto3.session.Session: S3 session.
+        """
+        try:
+            session = boto3.session.Session(
+                aws_access_key_id=s3_parameters["access-key"],
+                aws_secret_access_key=s3_parameters["secret-key"],
+                region_name=s3_parameters["region"],
+            )
+            session.resource("s3", endpoint_url=s3_parameters["endpoint"])
+            return session
+        except KeyError as e:
+            logger.warning("Missing required S3 parameter: %s", e)
+            return None
+        except ValueError as e:
+            logger.warning("Error creating resource: %s", e)
+            return None
+
+    def _retrieve_s3_parameters(self) -> Tuple[Dict, List[str]]:
```

Comment:
It's not clear to me why we return a tuple, I think I'd rather have 2 methods with independent purposes (`_get_missing_s3_parameters` and `_get_s3_parameters`) instead of having 1 method that does 2 things.

Reply:
Makes sense.
✅